### PR TITLE
feat(controller): Key-only artifacts. Fixes #3184

### LINKED ISSUE #3184: Define `artifactRepositoryRef` only once in spec
At the moment each new workflow must redefine an artifact repository:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: input-from-s3-

spec:
  entrypoint: input-from-s3
  serviceAccountName: argo
  templates:
  - name: input-from-s3
    container:
      image: alpine:3.12
      command: [sh, -c]
      args: ["cat /inputs/inputs.txt"]
    inputs:
      artifacts:
        - name: test-input
          path: /inputs
          s3:
            endpoint: s3.amazonaws.com
            region: us-east-2
            bucket: test-bucket
            key: inputs
            accessKeySecret:
              name: argo-s3-credentials
              key: accessKey
            secretKeySecret:
              name: argo-s3-credentials
              key: secretKey
```

Is that possible to define `s3` once and reuse it in different workflows? Something like that:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ArtifactRepo
metadata:
  name: my-input-artifact
spec:
  type: s3
  config:
    endpoint: s3.amazonaws.com
    region: us-east-1
    bucket: test-hadrien
    key: inputs
    accessKeySecret:
      name: argo-s3-credentials
      key: accessKey
    secretKeySecret:
      name: argo-s3-credentials
      key: secretKey
---
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: input-from-s3-

spec:
  entrypoint: input-from-s3
  serviceAccountName: argo
  templates:
  - name: input-from-s3
    container:
      image: alpine:3.12
      command: [sh, -c]
      args: ["cat /inputs/inputs.txt"]
    inputs:
      artifacts:
        - name: test-input
          path: /inputs
          artifactRepo: my-input-artifact
```

Or maybe this is already possible using workflow templates? If that's the case could you provide an example?

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 

Fixes #3184

## Goals

* Make it artifacts easier to work with.

Side-effects:

* Typically remove artifact information from `NodeStatus` and therefore reduces node status size, reducing etcd resource size. Node status is an implementation detail and uses SHOULD NOT read from it. Some maybe.
* Makes build caching trivially easy (CI use-case).
* Allows teams to configure artifact repositories on a per namespace basis.

### Refactoring

Basically pushing logic out of the controller and executor into `config` and `v1alpha1` packages where it is easy to test.

### Namespace-Level Default Artifact Repository

The ability to create a config map in a namespace and have that used automatically for repository configuration rather than having to make it explicit in every workflow.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: artifact-repositories
  annotations:
    workflows.argoproj.io/default-artifact-repository: default-v1
data:
  default-v1: |
    archiveLogs: true
    s3:
      bucket: my-bucket
      endpoint: minio:9000
      insecure: true
      accessKeySecret:
        name: my-minio-cred
        key: accesskey
      secretKeySecret:
        name: my-minio-cred
        key: secretkey
```

### Key-only Artifacts

This PR introduces the capability to specify artifacts only using the key - i.e. you do not need to specify the whole location.

```yaml
metadata:
  name: artifact-repository-ref
spec:
  entrypoint: main
  templates:
    - name: main
      dag:
        tasks:
          - name: generate
            template: generate
          - name: consume
            template: consume
            dependencies:
              - generate
    - name: generate
      container:
        image: argoproj/argosay:v2
        args:
          - echo
          - hello
          - /mnt/file
      outputs:
        artifacts:
          - name: file
            path: /mnt/file
            s3:
              key: my-file
    - name: consume
      container:
        image: argoproj/argosay:v2
        args:
          - cat
          - /tmp/file
      inputs:
        artifacts:
          - name: file
            path: /tmp/file
            s3:
              key: my-file
```

This is intended to work with artifact repository ref. Should it be allowed with the default repo?

See also: #3061 (#3063

### CODE REVIEW

File: test/e2e/testdata/artifact-repository-ref.yaml [hunk 1]
```diff
@@ -3,25 +3,43 @@ kind: Workflow
...
-      inputs:
+          - /mnt/file
+      outputs:
         artifacts:
           - name: file
-            path: /tmp/file
-            optional: true
-      outputs:
+            path: /mnt/file
+            s3:
```
Comment (Ark-kun) [inline test/e2e/testdata/artifact-repository-ref.yaml:32]: I wonder whether it's possible to simplify this further to a common `uri` field:
```
    uri: s3://my-bucket/my-file
```

Comment (alexec) [reply test/e2e/testdata/artifact-repository-ref.yaml:32]: something like go-getter?

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -3033,31 +3037,6 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
 	return false, nil
 }
 
-func (woc *wfOperationCtx) getArtifactRepositoryByRef(arRef *wfv1.ArtifactRepositoryRef) (*config.ArtifactRepository, error) {
```
Comment (alexec) [inline workflow/controller/operator.go:3036]: moved and re-written in 'artifactrepositories` package.

File: workflow/controller/scope.go [hunk 3]
```diff
@@ -92,47 +90,7 @@ func (s *wfScope) resolveArtifact(v string, subPath string) (*wfv1.Artifact, err
 
 		// Copy resolved artifact pointer before adding subpath
 		copyArt := valArt.DeepCopy()
-
-		locationType := copyArt.ArtifactLocation.GetType()
```
Comment (alexec) [inline workflow/controller/scope.go:96]: refactored into `wfv1.ArtifactLocationType.AppendToKey`

File: workflow/executor/executor.go [hunk 4]
```diff
@@ -294,57 +297,35 @@ func (we *WorkflowExecutor) saveArtifact(mainCtrID string, art *wfv1.Artifact) e
...
+}
+
+// fileBase is probably path.Base(filePath), but can be something else
+func (we *WorkflowExecutor) saveArtifactFromFile(art *wfv1.Artifact, fileName, localArtPath string) error {
+	if !art.HasKey() {
+		key, err := we.Template.ArchiveLocation.GetKey()
+		if err != nil {
+			return err
 		}
-		if we.Template.ArchiveLocation.S3 != nil {
```
Comment (alexec) [inline workflow/executor/executor.go:303]: refactored into `ArtifactLocationType.GetKey`, `SetType` and `SetKey`

File: workflow/executor/executor.go [hunk 5]
```diff
@@ -549,48 +530,12 @@ func (we *WorkflowExecutor) SaveLogs() (*wfv1.Artifact, error) {
 	if err != nil {
 		return nil, err
 	}
-	art := wfv1.Artifact{
```
Comment (alexec) [inline workflow/executor/executor.go:552]: refactored into `ArticactType.SetKey` etc

File: workflow/controller/workflowpod.go [hunk 6]
```diff
@@ -927,101 +923,22 @@ func addOutputArtifactsVolumes(pod *apiv1.Pod, tmpl *wfv1.Template) {
...
-		woc.log.Debugf("archive location unnecessary")
-		return nil
-	}
-	tmpl.ArchiveLocation = &wfv1.ArtifactLocation{
-		ArchiveLogs: woc.artifactRepository.ArchiveLogs,
-	}
-	// artifact location is defaulted using the following formula:
-	// <worflow_name>/<pod_name>/<artifact_name>.tgz
-	// (e.g. myworkflowartifacts/argo-wf-fhljp/argo-wf-fhljp-123291312382/src.tgz)
-	if s3Location := woc.artifactRepository.S3; s3Location != nil {
```
Comment (alexec) [inline workflow/controller/workflowpod.go:963]: refactored into `ArtifactRepositoryType.IntoArtifactLocation`

**Comment by alexec** [comment]
@AntoineDao would you like to review this for impact on subPath please? Thank you.

**Comment by AntoineDao** [comment]
Will give it a look + try just now and get back to you. Cheers for the heads up.

File: workflow/controller/scope.go (hunk 3 above)
Comment (AntoineDao) [reply workflow/controller/scope.go:96]: Very clean refactor. Tested and worked on my end + makes for cleaner artifact resolution logic + unit tests are happy without changing them.

We're good from a "subPath" perspective.

**Review by AntoineDao — APPROVED** [review]

**Comment by alexec** [comment]
TestCLIWithServerSuite/TestResourceTemplateStopAndTerminate/ResourceTemplateStop

**Pull request closed without merging**